webrender/src/device/gfx/descriptor_set.rs

```diff
@@ -0,0 +1,167 @@
+];
+
+pub(super) const COMMON_SET_3: &'static [DescriptorSetLayoutBinding] = &[
+    // Locals
+    descriptor_set_layout_binding!(0, DT::UniformBuffer, SSF::VERTEX, false),
+];
+
+pub(super) const CLIP_SET_0: &'static [DescriptorSetLayoutBinding] = &[
+    // Dither
+    descriptor_set_layout_binding!(0, DT::SampledImage, SSF::ALL, false),
+    // RenderTasks
+    // TODO: ShaderStageFlags should be VERTEX only here
+    descriptor_set_layout_binding!(1, DT::SampledImage, SSF::ALL, false),
+    // GpuCache
+    descriptor_set_layout_binding!(2, DT::SampledImage, SSF::ALL, false),
+    // TransformPalette
+    // TODO: ShaderStageFlags should be VERTEX only here
+    descriptor_set_layout_binding!(3, DT::SampledImage, SSF::ALL, false),
+];
+
+pub(super) const CLIP_SET_1: &'static [DescriptorSetLayoutBinding] = &[
+    // Color0 sampler
+    descriptor_set_layout_binding!(12, DT::Sampler, SSF::ALL, false),
+    // Color1 sampler
+    descriptor_set_layout_binding!(13, DT::Sampler, SSF::ALL, false),
+    // Color2 sampler
+    descriptor_set_layout_binding!(14, DT::Sampler, SSF::ALL, false),
+    // Dither sampler
+    descriptor_set_layout_binding!(0, DT::Sampler, SSF::ALL, false),
+    // RenderTasks sampler
+    // TODO: ShaderStageFlags should be VERTEX only here
+    descriptor_set_layout_binding!(1, DT::Sampler, SSF::ALL, false),
+    // GpuCache sampler
+    descriptor_set_layout_binding!(2, DT::Sampler, SSF::ALL, false),
+    // TransformPalette sampler
+    // TODO: ShaderStageFlags should be VERTEX only here
+    descriptor_set_layout_binding!(3, DT::Sampler, SSF::ALL, false),
+];
+
+pub(super) const CLIP_SET_2: &'static [DescriptorSetLayoutBinding] = DEFAULT_SET_2;
+
+pub(super) const PRIMITIVE_SET_0: &'static [DescriptorSetLayoutBinding] = &[
+    // Dither
+    descriptor_set_layout_binding!(0, DT::SampledImage, SSF::ALL, false),
+    // RenderTasks
+    // TODO: ShaderStageFlags should be VERTEX only here
+    descriptor_set_layout_binding!(1, DT::SampledImage, SSF::ALL, false),
+    // GpuCache
+    descriptor_set_layout_binding!(2, DT::SampledImage, SSF::ALL, false),
+    // TransformPalette
+    // TODO: ShaderStageFlags should be VERTEX only here
+    descriptor_set_layout_binding!(3, DT::SampledImage, SSF::ALL, false),
+    // PrimitiveHeadersF
+    // TODO: ShaderStageFlags should be VERTEX only here
+    descriptor_set_layout_binding!(4, DT::SampledImage, SSF::ALL, false),
+    // PrimitiveHeadersI
+    // TODO: ShaderStageFlags should be VERTEX only here
+    descriptor_set_layout_binding!(5, DT::SampledImage, SSF::ALL, false),
+];
+
+pub(super) const PRIMITIVE_SET_1: &'static [DescriptorSetLayoutBinding] = &[
+    // Color0 sampler
+    descriptor_set_layout_binding!(12, DT::Sampler, SSF::ALL, false),
+    // Color1 sampler
+    descriptor_set_layout_binding!(13, DT::Sampler, SSF::ALL, false),
+    // Color2 sampler
+    descriptor_set_layout_binding!(14, DT::Sampler, SSF::ALL, false),
+    // PrevPassAlpha sampler
```

Comment:
Samplers of per-pass data should go to SET 0. Pretty sure you can get away with a single sampler to be used for both color and alpha, no need for 2.